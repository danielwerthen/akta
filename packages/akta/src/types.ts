import { Observable } from 'rxjs';

export type AktaElement<
  P extends { [key: string]: unknown } = { [key: string]: unknown }
> = {
  $$type: Symbol;
  type: string | AktaComponent<P>;
  props: P;
  key: string | undefined;
};

export type ObservableElements =
  | Observable<AktaAllElements>
  | Generator<AktaAllElements, AktaAllElements>
  | AsyncGenerator<AktaAllElements, AktaAllElements>;

export type AktaAllElements = AktaElement | ObservableElements;

export type AktaNode = AktaAllElements | string | null | AktaNode[];

export const AktaElementType = Symbol('Akta element');

export function isAktaElement(node: unknown): node is AktaElement {
  if (typeof node === 'object') {
    return (node as AktaElement)?.$$type === AktaElementType;
  }
  return false;
}

export interface ElementProperties {}

export type AktaComponent<PROPS extends {}> = (
  props: PROPS
) => AktaAllElements | string | null;

export function AktaPreparedComponent(_props: {
  children: Observable<HTMLElement>;
}) {
  return null;
}
