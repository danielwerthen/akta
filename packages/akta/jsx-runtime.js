export { Fragment, jsx, jsxs } from './dist';
