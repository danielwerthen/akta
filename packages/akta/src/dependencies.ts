import { BehaviorSubject, firstValueFrom, Subject, switchMap } from 'rxjs';
import { createDependency, DependencyMap } from './dependency-map';
import { AllElements, isEvent } from './element-ops';
import { lazy } from './lazy-function';
import MetaObject, { MethodMissing } from './meta-object';
import { createSyncContext } from './sync-context';
import { AktaNode, ElementProperties } from './types';

export const dependecyContext = createSyncContext<DependencyMap>();

export const continuationDependency = createDependency(lazy());
export type TeardownFunction = () => void;
export const teardownDependency = createDependency<null | TeardownFunction[]>(
  null
);

export const elementsDependency = createDependency(new AllElements());

export function useNext<T extends AktaNode, RETURN extends AktaNode, NEXT>(
  _input?: () => Generator<T, RETURN, NEXT> | AsyncGenerator<T, RETURN, NEXT>
): (input: NEXT) => Promise<void> {
  const ctx = dependecyContext.getContext();
  return (input: NEXT) => {
    return firstValueFrom(
      ctx(continuationDependency).pipe(switchMap(next => next(input)))
    ).then(() => void 0);
  };
}

export function useTeardown(fn: TeardownFunction) {
  const ctx = dependecyContext.getContext();
  ctx.provide(teardownDependency, fns => {
    if (!fns) {
      return [fn];
    }
    fns.push(fn);
    return fns;
  });
}

export class PropertiesBase extends MetaObject<
  | ((factory?: () => Subject<unknown>) => Subject<unknown>)
  | ((start?: unknown) => Subject<unknown>)
> {
  [MethodMissing](key: string) {
    if (isEvent(key)) {
      return (factory?: () => Subject<unknown>) =>
        factory ? factory() : new Subject();
    } else {
      return (initialValue?: unknown) => {
        if (initialValue === undefined) {
          return new Subject();
        } else {
          return new BehaviorSubject(initialValue);
        }
      };
    }
  }
}

const propertiesBase = new PropertiesBase();
export class ElementPropertiesBase extends MetaObject<PropertiesBase> {
  [MethodMissing](_key: string) {
    return propertiesBase;
  }
}

export const Props: ElementProperties = (new ElementPropertiesBase() as unknown) as ElementProperties;
