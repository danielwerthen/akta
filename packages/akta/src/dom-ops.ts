import {
  combineLatest,
  firstValueFrom,
  from,
  isObservable,
  Observable,
  of,
} from 'rxjs';
import { filter, finalize, map, mapTo, switchMap, tap } from 'rxjs/operators';
import {
  continuationDependency,
  dependecyContext,
  teardownDependency,
} from './dependencies';
import { createDependencyMap, DependencyMap } from './dependency-map';
import { AllElements, mountElement, unmountElement } from './element-ops';
import { lazy } from './lazy-function';
import {
  AktaAllElements,
  AktaComponent,
  AktaElement,
  isAktaElement,
} from './types';

type AktaContext = {
  intrinsic: AllElements;
  dependencies: DependencyMap;
};

function applyProp(
  element: HTMLElement,
  type: string,
  key: string,
  value: unknown,
  ctx: AktaContext
): Observable<unknown> | void {
  return ctx.intrinsic[type][key](element, value);
}

function onlyFirst(_value: unknown, idx: number) {
  return idx === 0;
}

const emptyNode = () => document.createTextNode('');

function isGenerator(
  obj: unknown
): obj is Generator<any, any, any> | AsyncGenerator<any, any, any> {
  if (typeof obj === 'object') {
    return typeof (obj as Generator<any, any, any>).next === 'function';
  }
  return false;
}

export function callComponent<PROPS>(
  component: AktaComponent<PROPS>,
  props: PROPS,
  parentDeps: DependencyMap
): [AktaAllElements, DependencyMap] {
  const deps = parentDeps.branch();
  const element = dependecyContext.setContext(() => {
    return component(props);
  }, deps);
  if (isObservable(element)) {
    return [element, deps];
  }
  if (isGenerator(element)) {
    const continuation = lazy();
    deps.provide(continuationDependency, continuation);
    const generated = dependecyContext.setContext(() => {
      return element.next();
    }, deps);
    const observable = new Observable<AktaAllElements>(subscriber => {
      Promise.resolve(generated).then(({ value, done }) => {
        subscriber.next(value);
        if (done) {
          subscriber.complete();
          return;
        }
        let isComplete = false;
        continuation.define(input => {
          if (isComplete) {
            return of();
          }
          const generated = dependecyContext.setContext(() => {
            return element.next(input);
          }, deps);
          return from(Promise.resolve(generated)).pipe(
            map(({ value, done }) => {
              if (isComplete) {
                return;
              }
              subscriber.next(value);
              if (done) {
                isComplete = true;
                subscriber.complete();
                return;
              }
            })
          );
        });
      });
    });
    return [observable, deps];
  }
  return [element, deps];
}

function applyChildren(
  children: AktaAllElements | AktaAllElements[],
  parent: HTMLElement,
  ctx: AktaContext
): Observable<unknown> {
  if (Array.isArray(children)) {
    const observables = children
      .map(child => produceElements(child, ctx))
      .map(observedChild => {
        let oldNode: HTMLElement | Text | undefined;
        return observedChild.pipe(
          tap(newNode => {
            if (oldNode) {
              unmountElement(oldNode);
              mountElement(newNode);
              parent.replaceChild(newNode, oldNode);
            } else {
              parent.appendChild(newNode);
              mountElement(newNode);
            }
            oldNode = newNode;
          }),
          filter(onlyFirst)
        );
      });
    return combineLatest(observables).pipe(mapTo(void 0));
  }
  let oldNode: HTMLElement | Text;
  return produceElements(children, ctx).pipe(
    tap(newNode => {
      if (oldNode) {
        parent.replaceChild(newNode, oldNode);
        mountElement(newNode);
        unmountElement(oldNode);
      } else {
        parent.appendChild(newNode);
        mountElement(newNode);
      }
      oldNode = newNode;
    }),
    filter(onlyFirst)
  );
}

function produceElement(
  node: AktaElement,
  ctx: AktaContext
): Observable<HTMLElement | Text> {
  const { type, props } = node;
  if (typeof type === 'string') {
    const element = document.createElement(type);
    const observables: Observable<unknown>[] = [];
    for (var key in props) {
      const observable =
        key === 'children'
          ? applyChildren(props[key] as AktaAllElements, element, ctx)
          : applyProp(element, type, key, props[key], ctx);
      if (observable) {
        observables.push(observable.pipe(filter(onlyFirst)));
      }
    }
    if (observables.length < 1) {
      return of(element);
    }
    return combineLatest(observables).pipe(mapTo(element));
  } else if (typeof type === 'function') {
    const [element, dependencies] = callComponent(
      type,
      props,
      ctx.dependencies
    );

    const teardown = dependencies.get(teardownDependency);
    const output = produceElements(element, { ...ctx, dependencies });
    if (teardown) {
      return output.pipe(finalize(teardown));
    }
    return output;
  } else {
    return of(emptyNode());
  }
}

function isPromise(obj: unknown): obj is Promise<unknown> {
  return typeof (obj as Promise<unknown>)?.then === 'function';
}

function produceElements(
  node: AktaAllElements,
  ctx: AktaContext
): Observable<HTMLElement | Text> {
  if (isObservable(node)) {
    return node.pipe(
      switchMap(innerNode => {
        return produceElements(innerNode, ctx);
      })
    );
  } else if (isPromise(node)) {
    return from(node).pipe(
      switchMap((innerNode: AktaAllElements) => {
        return produceElements(innerNode, ctx);
      })
    );
  } else if (isAktaElement(node)) {
    return produceElement(node, ctx);
  } else {
    return of(document.createTextNode(node.toString()));
  }
}

export function prepare(
  element: AktaAllElements,
  root: HTMLElement
): Promise<void> {
  const dependencies = dependecyContext.getContext();
  const ctx = {
    dependencies,
    intrinsic: new AllElements(),
  };
  return firstValueFrom(applyChildren(element, root, ctx).pipe(mapTo(void 0)));
}

export function mount(element: AktaAllElements, root: HTMLElement) {
  const ctx = {
    dependencies: createDependencyMap(),
    intrinsic: new AllElements(),
  };
  const sub = applyChildren(element, root, ctx).subscribe();

  return () => sub.unsubscribe();
}
