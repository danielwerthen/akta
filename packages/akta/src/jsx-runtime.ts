// Based on https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/314d0c3cf25e1762525a3790b4609eb19fddadf9/types/react/index.d.ts

import { AktaElement, AktaElementType } from './types';

export function jsx<
  P extends {},
  T extends string | ((props: P) => AktaElement<P, any>)
>(type: T, props: P, maybeKey?: string): AktaElement<P, T> {
  return {
    $$type: AktaElementType,
    type,
    props,
    key: maybeKey,
  };
}

export const jsxs = jsx;
