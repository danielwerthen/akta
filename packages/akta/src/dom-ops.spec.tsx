import { firstValueFrom, from, Observable, Subject, toArray } from 'rxjs';
import { useTeardown } from './dependencies';
import { mount, prepare } from './dom-ops';
import { jsx } from './jsx-runtime';
import { AktaNode } from './types';

describe('DOM OPS', () => {
  it('Observable child nodes are rendered correctly', async () => {
    const root = document.createElement('div');
    const value = new Observable(sub => {
      setTimeout((item: string) => sub.next(item), 10, 'foobar');
    });
    const promise = prepare(
      jsx('p', {
        children: ['top', from(['immediate', 'immediate2']), value, 'bottom'],
      }),
      root
    );
    expect(root).toMatchSnapshot();
    await promise;
    await new Promise(res => setTimeout(res, 100));
    expect(root).toMatchSnapshot();
  });
  it('should render within range of comparison implementation', async () => {
    const root = document.createElement('div');
    const array = new Array(100).fill(0);
    const value = array.map(() =>
      jsx('span', { children: 'word', className: 'foobar' })
    );
    const t0 = performance.now();
    await prepare(jsx('p', { children: value }), root);
    function update(span: HTMLSpanElement) {
      span.classList.add('foobar');
      span.innerText = 'word';
    }
    const t1 = performance.now();
    const root2 = document.createElement('div');
    array.forEach(() => {
      const span = document.createElement('span');
      update(span);
      root2.appendChild(span);
    });
    const t2 = performance.now();

    const ratio = Math.round((100 * (t1 - t0)) / (t2 - t1)) / 100;
    expect(ratio).toBeLessThan(10);
  });

  it('Array nodes are rendered correctly', async () => {
    const root = document.createElement('div');
    const value = ['test', 'more'];
    await prepare(jsx('p', { children: value }), root);
    expect(root).toMatchSnapshot();
  });

  it('mount and unmount is called at the right time', async () => {
    const root = document.createElement('div');
    const value = new Subject<AktaNode>();
    const unsub = mount(jsx('p', { children: value }), root);
    const onMount = new Subject<unknown>();
    const onUnmount = new Subject<unknown>();
    const mounts = firstValueFrom(onMount.pipe(toArray()));
    const unmounts = firstValueFrom(onUnmount.pipe(toArray()));
    value.next(jsx('div', { onMount, onUnmount, children: 'first' }));
    await new Promise(res => setTimeout(res, 10));
    value.next(jsx('div', { children: 'final' }));
    await new Promise(res => setTimeout(res, 10));
    unsub();
    onMount.complete();
    onUnmount.complete();
    expect(root).toMatchSnapshot();
    expect((await mounts).length).toBe(1);
    expect((await unmounts).length).toBe(1);
  });

  it('deep mount and unmount is called at the right time', async () => {
    const root = document.createElement('div');
    const value = new Subject<AktaNode>();
    const unsub = mount(jsx('p', { children: value }), root);
    const onMount = new Subject<unknown>();
    const onUnmount = new Subject<unknown>();
    const mounts = firstValueFrom(onMount.pipe(toArray()));
    const unmounts = firstValueFrom(onUnmount.pipe(toArray()));
    value.next(
      jsx('div', {
        children: jsx('div', {
          onMount,
          onUnmount,
          children: jsx('div', { onMount, onUnmount, children: 'first' }),
        }),
      })
    );
    await new Promise(res => setTimeout(res, 10));
    value.next(jsx('div', { children: 'final' }));
    await new Promise(res => setTimeout(res, 10));
    unsub();
    onMount.complete();
    onUnmount.complete();
    expect(root).toMatchSnapshot();
    expect((await mounts).length).toBe(2);
    expect((await unmounts).length).toBe(2);
  });
  it('should handle teardown', async () => {
    const root = document.createElement('div');
    const value = new Subject<AktaNode>();
    const teardown = jest.fn();
    function Comp() {
      useTeardown(teardown);
      return jsx('p', { children: 'Data' });
    }
    const unsub = mount(jsx('div', { children: value }), root);
    value.next(jsx(Comp, {}));
    expect(teardown.mock.calls.length).toBe(0);
    value.next(jsx(Comp, {}));
    expect(teardown.mock.calls.length).toBe(1);
    value.next(jsx(Comp, {}));
    expect(teardown.mock.calls.length).toBe(2);
    unsub();
    expect(teardown.mock.calls.length).toBe(3);
    expect(root).toMatchSnapshot();
  });
});
