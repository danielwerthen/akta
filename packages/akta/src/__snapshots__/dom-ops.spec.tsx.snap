// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOM OPS Array nodes are rendered correctly 1`] = `
<div>
  <p>
    test
    more
  </p>
</div>
`;

exports[`DOM OPS Observable child nodes are rendered correctly 1`] = `<div />`;

exports[`DOM OPS Observable child nodes are rendered correctly 2`] = `
<div>
  <p>
    top
    immediate2
    foobar
    bottom
  </p>
</div>
`;

exports[`DOM OPS deep mount and unmount is called at the right time 1`] = `
<div>
  <p>
    <div>
      final
    </div>
  </p>
</div>
`;

exports[`DOM OPS mount and unmount is called at the right time 1`] = `
<div>
  <p>
    <div>
      final
    </div>
  </p>
</div>
`;

exports[`DOM OPS should handle observable child 1`] = `
<div>
  <p>
    Daniel
  </p>
</div>
`;

exports[`DOM OPS should handle prepare 1`] = `
<div>
  <p>
    Loading
  </p>
</div>
`;

exports[`DOM OPS should handle prepare 2`] = `
<div>
  <p>
    <span>
      Finally
    </span>
  </p>
</div>
`;

exports[`DOM OPS should handle teardown 1`] = `
<div>
  <div>
    <p>
      Data3
    </p>
  </div>
</div>
`;
